test/unit/link_checker_test.rb

```diff
@@ -0,0 +1,94 @@
+require "test_helper"
+require "gds_api/test_helpers/link_checker_api"
+
+class LinkCheckerTest < ActiveSupport::TestCase
+  include GdsApi::TestHelpers::LinkCheckerApi
+
+  def url
+    @url ||= "http://example.com"
+  end
+
+  def text
+    %(There is a link <a href="#{url}">here</a>)
+  end
+
+  def data
+    @data ||= text
+  end
+
+  def link_checker
+    @link_checker ||= LinkChecker.new(data)
+  end
+
+  def fixtures_path
+    Rails.root.join "test/fixtures/smart_answer_flows/flow-with-links"
+  end
+
+  context "#urls" do
+    should "return set of urls in text" do
+      assert_equal Set[url], link_checker.urls
+    end
+
+    context "when text in an array" do
+      should "return set of urls in text" do
+        @data = [text]
+        assert_equal Set[url], link_checker.urls
+      end
+    end
+
+    context "when text contains no links" do
+      should "return an empty set" do
+        @data = "No links"
+        assert_equal Set[], link_checker.urls
+      end
+    end
+
+    context "when data missing" do
+      should "return an empty set" do
+        @link_checker = LinkChecker.new(nil)
+        assert_equal Set[], link_checker.urls
+      end
+    end
+  end
+
+  context "#report" do
+    should "generate a report" do
+      stub_link_checker_api_create_batch(
+        uris: [url], checked_within: 5, status: :completed,
+      )
+      assert_equal url, link_checker.report.links.first.uri
+    end
+  end
+
+  context "#report_text" do
+    should "output summary report as text" do
+      stub_link_checker_api_create_batch(
+        uris: [url], checked_within: 5, status: :completed,
+      )
+      assert_include link_checker.report_text, "LINKS TO CHECK 1"
+      assert_include link_checker.report_text, "Broken links"
+      assert_include link_checker.report_text, "Cautious links"
+    end
+  end
+
+  context ".for_files_at" do
+    should "run report for files at path" do
+      link_checker = LinkChecker.for_files_at(fixtures_path)
+      assert_equal 3, link_checker.urls.length
+    end
+
+    should "extract string delimited urls" do
+      link_checker = LinkChecker.for_files_at(fixtures_path)
+      assert link_checker.urls.include?("http://example.com/one")
+    end
+
+    should "extract url from govspeak markdown" do
+      link_checker = LinkChecker.for_files_at(fixtures_path)
+      assert link_checker.urls.include?("https://example.com/two")
+    end
+  end
```

Comment:
I feel like here, we've got a weird hybrid test of integration vs unit test. These tests are covering a lot of functionality including the ability to read in from files. i.e. can we read in a file path, can we read multiple files etc... The testing of the actual text parsing for URLs seems quite cumbersome - and would be easier if there was a discrete method for extract URLs from a string.

Reply:
There is a discrete method for extracting URLs from a string. Its [`URI.extract`](https://ruby-doc.org/stdlib-2.7.0/libdoc/uri/rdoc/URI/RFC2396_Parser.html#method-i-extract) which is part of the standard Ruby library.

Reply:
Yup - so there's a scoped functionality around being able to extract urls from text - i.e. making sure we are calling url.extract and doing bracket removal. But that functionality isn't part of the `for_files_at` method we are testing here. We should be asserting that we are getting a LinkChecker object with `texts` attribute set to the correct value. We are testing more than we need too which will make them fragile - e.g. if the `urls` breaks, then all these tests would fails, which isn't necessarily indicative of an issue with `for_fails_at` method